src/service/VerifyService.ts

```diff
@@ -41,91 +40,100 @@ const defaultExpectedVersions: ExpectedVersions = {
-        if (this.expectedVersions.symbolBootstrap) lines.push(await this.testSymbolBootstrap(this.expectedVersions.symbolBootstrap));
-        if (!docker.recommendation) lines.push(await this.testDockerRun());
-        if (!BootstrapUtils.isWindows()) {
-            lines.push(await this.testSudo());
-        }
-
-        return { lines, platform };
-    }
-
-    public logReport(report: VerifyReport): void {
-        this.logger.info(`OS: ${report.platform}`);
-        report.lines.forEach((line) => {
-            if (line.recommendation) {
-                this.logger.error(`${line.header}  - Error! - ${line.message} - ${line.recommendation}`);
-            } else {
-                this.logger.info(`${line.header} - OK! - ${line.message}`);
-            }
-        });
-    }
-
-    public validateReport(report: VerifyReport): void {
-        const errors = report.lines.filter((r) => r.recommendation);
-        if (errors.length) {
-            throw new Error(
-                'There has been an error. Check the report: \n' +
-                    errors.map((line) => ` - ${line.header}  - Error! - ${line.message} - ${line.recommendation}`).join('\n'),
-            );
-        }
-    }
+export interface VerifyAction {
+    shouldRun(lines: ReportLine[]): Promise<boolean>;
+    verify(): Promise<ReportLine>;
+}
 
+export class AppVersionService {
+    public static readonly semverOptions = { loose: true };
+    constructor(private readonly logger: Logger) {}
     public loadVersion(text: string): string | undefined {
         return text
             .replace(',', '')
             .split(' ')
             .map((word) => {
-                const coerce = semver.coerce(word.trim(), this.semverOptions);
+                const coerce = semver.coerce(word.trim(), AppVersionService.semverOptions);
                 return coerce?.raw;
             })
             .find((a) => a)
             ?.trim();
     }
 
-    public async testNodeJs(expectedVersion: string): Promise<ReportLine> {
-        const header = 'NodeVersion';
-        const recommendationUrl = `https://nodejs.org/en/download/package-manager/`;
-        const output = process.versions.node;
-        return this.verifyInstalledApp(async () => output, header, expectedVersion, recommendationUrl);
+    public async loadVersionFromCommand(command: string): Promise<string | undefined> {
+        return this.loadVersion((await BootstrapUtils.exec(this.logger, command)).stdout.trim());
     }
 
-    public async testDocker(expectedVersion: string): Promise<ReportLine> {
-        const header = 'Docker Version';
-        const command = 'docker --version';
-        const recommendationUrl = `https://docs.docker.com/get-docker/`;
-        return this.verifyInstalledApp(() => this.loadVersionFromCommand(command), header, expectedVersion, recommendationUrl);
+    public async verifyInstalledApp(
+        versionLoader: () => Promise<string | undefined>,
```

Comment:
when can this return undefined? 
in code below, it looks like it throws on failure.

Reply:
The exception is for misconfigured actions.  A well-configured action may have the issue that the command is not returning any version (so the possible undefinded)

https://github.com/fboucquez/symbol-bootstrap/blob/cbe72d35b05b6d692ba20041a3d08afb65d812c1/src/service/VerifyService.ts#L63